crates/noirc_evaluator/src/ssa/acir_gen/operations/sort.rs

```diff
@@ -40,10 +61,17 @@ pub fn permutation_layer(
     }
     let n1 = n / 2;
     let mut conf = Vec::new();
+
     // witness for the input switches
-    for _ in 0..n1 {
-        conf.push(evaluator.add_witness_to_cs());
+    #[allow(clippy::needless_range_loop)]
+    for i in 0..n1 {
+        if generate_witness {
+            conf.push(evaluator.add_witness_to_cs());
+        } else {
+            conf.push(bits[i]);
+        }
```

Comment:
```rs
let conf = vecmap(bits.iter().take(n1), |bit| {
    if generate_witness {
        evaluator.add_witness_to_cs()
    } else {
        *bit
    }
});
```
To get rid of the clippy warning

Reply:
The clippy warning is wrong so it is fine to disable it.
You suggestion removes the warning but yields incorrect code.

Reply:
- I don't see why the warning is incorrect. It seems to be suggesting to loop over `bits.iter().take(n1)` instead.
- Why is my suggestion incorrect?

Reply:
For the same reason the warning is not correct. It obviously does not work when bits is empty.

Reply:
Ah, I was not expecting if `generate_witness` is true then bits may be empty. Is bits always completely empty if `generate_witness` is true? If so perhaps we could represent it more directly e.g. via `bits: Option<&[Witness]>`.

For now, a more direct conversion to get rid of the clippy warning would be:
```rs
let conf = vecmap(0..n1, |i| { 
    if generate_witness { 
        evaluator.add_witness_to_cs()
    } else {
        bits[i]
    }
});
```

Reply:
If generate_witness is true, the function does not use bits, so it can be anything. In practice it is empty because we don't have them and that's why we want to generate them. Anyways, since the clippy warning is wrong, I don't see why I should change the code, because of a wrong warning.

Reply:
We should change the code because the suggestion is more concise and clearer with the additional benefit of no longer needing to silence the warning.

Reply:
The suggestion is not more concise (loop and vecmap are 7 lines) and it does exactly the same thing (iterating over a range). The difference is that the linter is not smart enough to see this, which proves it is not as straightforward than using a simple loop.
Anyways I don't care much and I would like this to be merged because other PRs depend on it, so I did use the suggestion.

Reply:
Thank you for the change. The suggestion is more concise, using `wc` on both
```rs
let mut conf = Vec::new();
for i in 0..n1 {
    if generate_witness {
        conf.push(evaluator.add_witness_to_cs());
    } else {
        conf.push(bits[i]);
    }
}
```
and
```rs
let mut conf = iter_extended::vecmap(0..n1, |i| {
    if generate_witness {
        evaluator.add_witness_to_cs()
    } else {
        bits[i]
    }
});
```
It says the former is 8 lines, 20 words, and 169 characters, while the later is 7 lines, 17 words, and 153 characters. This is with the clippy annotation and extra line of whitespace removed as well. It is nothing major as you say, but I still count it as a readability improvement.